### PR TITLE
Allow empty tiered replicants map for load rules

### PR DESCRIPTION
**Pull request by adarshsanjeev**
This PR makes some changes around the allowed tiered replicants for load rules. It allows values empty tiered replicant in load rules, which would translate to the segment being marked as used, but not loaded on any historical for example:

For backwards compatibility, a new flag "allowEmptyTieredReplicants" is also added. If the flag is false or missing, the old behaviour of converting an empty map to the default load rule is maintained. If it is true, which would be the case going forward, empty load rules are allowed.

```
{
    "interval": "2019-08-25T12:00:00.000Z/2019-08-26T00:00:00.000Z",
    "type": "loadByInterval",
    "allowEmptyTieredReplicants": true
}
```

The above load rule would not be loaded on the historical as per the new behaviour.

Exceptions thrown during validation are returned to the user.
Screenshot 2023-06-20 at 8 39 57 AM

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by kfaraz — COMMENTED** [review]
I think this change is backwards incompatible. There must be several clusters using `null` which internally gets converted to the mapping `_default_tier => 2`. You are now converting these to the mapping `_default_tier => 0` in the `LoadRule.run()` method. This will cause unexpected behaviour in all such clusters.

I can think of some alternatives:

#### Option 1 (preferable)
- Continue to treat `null` as a special and convert it to `_default_tier => 2` as before
- Also allow empty in the `LoadRule.run()`. It would automatically translate to mean the segment need not be loaded anywhere.
- Allow `0` replicas to be specified for any tier for the same backward compatibility reasons. I don't think this affects the coordinator behaviour in any way.

The only con of this method is that `null` and `empty` have different meanings.

OR

#### Option 2 (not very user-friendly)
- Keep everything as is.
- We must specify mapping for at least one tier even it we don't want segment to be loaded anywhere.

#### Option 3 (reserve a special tier name)
- Similar to option 2 but use a special tier-name which can be mapped to 0 instead of an actual tier in the cluster.

@adarshsanjeev , @cryptoe , what do you think?

**Comment by adarshsanjeev** [comment]
I think Option 1 has the potential to be confusing. For example
```
  {
    "type": "loadByInterval",
    "tieredReplicants": {},
    "interval": "2010-01-01/2020-01-01"
  }
```
and 
```
  {
    "type": "loadByInterval",
    "interval": "2010-01-01/2020-01-01"
  }
```
would result in one loading the segments on historicals and one would not, which would end up being difficult to understand moving forward.

**Comment by cryptoe** [comment]
We could also include a new flag called `allowNullTieredReplicants` which is defaulted to `true` and does not change the current behavior ie we set the default tier rules in case `NullTieredReplicants` is null. 

If users set this to `false` the new behavior is enabled. The web-console starts setting this new flag for every request.  That would solve the backward compatibility piece. 
Its `Option 1` with a new flag. 

cc @kfaraz @adarshsanjeev

File: server/src/main/java/org/apache/druid/server/coordinator/rules/ForeverLoadRule.java
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/ForeverLoadRule.java:43]:
```suggestion
      @JsonProperty("useDefaultTierForNull") @Nullable Boolean useDefaultTierForNull
```

**Review by cryptoe — COMMENTED** [review]
Just suggested a new name.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalLoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalLoadRule.java:50]: You can use the utility method:
```suggestion
    this.useDefaultTierForNull = Configs.valueOrDefault(useDefaultTierForNull, true);
```

Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalLoadRule.java]: Thanks! This cleaned the code a lot.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalLoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/IntervalLoadRule.java]: Refactored this into a separate function instead, since `validateAndGetTieredReplicants` was starting to do two different things. Please let me know if this looks good.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java:56]: Since this is set via an API RulesResources#setDatasourceRules(). How would the exception thrown here be visible to end users. 

I guess we need to send a message back in the API response.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:51]: Maybe rename this to `handleNullTieredReplicants` or similar.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:50]:
```suggestion
  /**
   * Returns the given {@code tieredReplicants} map unchanged if it is non-null (including empty).
   * Returns the following default values if the given map is null.
   * <ul>
   * <li>If {@code useDefaultTierForNull} is true, returns a singleton map from {@link DruidServer#DEFAULT_TIER} to {@link DruidServer#DEFAULT_NUM_REPLICANTS}.</li>
   * <li>If {@code useDefaultTierForNull} is false, returns an empty map. This causes segments to have a replication factor of 0 and not get assigned to any historical.</li>
   * </ul>
   */
```

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java:56]: In master, the message is returned as a warning after submitting the load rule. This behaviour here would be the same, or did I misunderstand what you mean?

Screenshot 2023-06-20 at 8 39 57 AM

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:64]: Do we still need this validation? If there is a valid use-case for empty tiered replicants, I think we can just allow it regardless of the value of `useDefaultTierForNull`. Also `useDefaultTierForNull` is just meant to handle null values and using it here becomes confusing.

File: services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java
Comment (kfaraz) [inline services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java:112]: Maybe pass `true` explicitly in all tests where you want to retain the old behaviour. It would help you distinguish between tests for old and new behaviour.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java:215]: Use a static class-level field instead.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java:202]: Use `Assert.assertThrows` instead and verify the exception message too.

**Review by kfaraz — COMMENTED** [review]
Minor changes required.

File: services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java
Comment (adarshsanjeev) [reply services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java:112]: Would it make more sense to use null and false to differentiate old and new tests, since that is the real test for backward compatibility?

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java
Comment (cryptoe) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java:56]: Sure this looks good to me. We can share this screen shot as part of the PR description.

**Review by cryptoe — APPROVED** [review]
Overall lgtm!!
Will merge once @kfaraz gives his approval.

File: services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java
Comment (kfaraz) [reply services/src/test/java/org/apache/druid/server/router/CoordinatorRuleManagerTest.java:112]: Sure, that works too as long as there are at least a couple of tests for each of the 3 flavors: null (for backward compat), true and false.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:67]: Try using this exception in the next line instead:

```suggestion
      throw InvalidInput.exception("Number of replicas in tier [%s] is null", entry.getKey());
```

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:67]: Similarly for the case `entry.getValue() < 0`, use this instead:

```suggestion
      throw InvalidInput.exception("Number of replicas for tier [%s] must be positive. Found [%d].", entry.getKey(), entry.getValue());
```

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java]: Initially, this was not possible as the Coordinator code was breaking with an empty map in a few places. I observed that the segments were being dropped even if the rule was present. However, I think thanks to your coordinator changes, this has been fixed. Reverting the changes to this function as it seem to be not needed.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java
Comment (adarshsanjeev) [reply server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Changed

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java]: Moved here, I've kept the Jackson the same for now, but the code repetition has reduced a lot.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java]: Removed this for now, I guess that this would not break backward compatibility since we are relaxing the validations.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:84]: Changed these exception types.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:36]: Why do these fields need to be protected? They can remain private and the sub-classes can call the getters when needed.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:72]: I think both these methods `handleNull` and `validate` can be private now.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodLoadRule.java:76]: This method is also common across the 3 load rule implementations, could be moved into `LoadRule`.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java:66]: Nit: In test assertions, you can use the final message with values interpolated, so that the reader knows the message that is being verified.
```
Invalid number of replicas for tier [_default_tier]. Value [-1] must be positive.
```

File: server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java:63]: Nit: Can use `expectMessageContains` instead of a `Matcher.containsString`.

**Review by kfaraz — APPROVED** [review]
Thanks for the changes, @adarshsanjeev !

File: server/src/main/java/org/apache/druid/server/coordinator/rules/ForeverLoadRule.java
Comment (abhishekrb19) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/ForeverLoadRule.java:38]: Should we document this property along with an example where `tieredReplicants` is null: https://github.com/apache/druid/blob/master/docs/operations/rule-configuration.md

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (abhishekrb19) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:33]: Could use a javadoc update with the new property addition.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java
Comment (abhishekrb19) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/ForeverLoadRuleTest.java:123]: Thanks for these tests!

**Review by abhishekrb19 — COMMENTED** [review]
LGTM. Do we want to document this new parameter: https://github.com/apache/druid/blob/master/docs/operations/rule-configuration.md?

File: server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java
Comment (adarshsanjeev) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/LoadRule.java:33]: Added doc to the parameter as well. I think `validateTieredReplicants` already covers it.

File: docs/operations/rule-configuration.md
Comment (kfaraz) [inline docs/operations/rule-configuration.md:120]: Examples are not needed here, as this section is common to all load rules. You can either leave out the examples for now or just give one example in any one of the specific load rule implementations.

File: docs/operations/rule-configuration.md
Comment (kfaraz) [inline docs/operations/rule-configuration.md:118]: All load rules can have these properties:

|Property|Description|Required|Default value|
|---------|-----------|---------|-------------|
| `tieredReplicants`| Map from tier names to the respective number of segment replicas to be loaded on those tiers. The number of replicas for each tier must be either 0 or a positive integer.| No | When `useDefaultTierForNull` is `true`, the default value is `{"_default_tier": 2}` i.e. 2 replicas to be loaded on the `_default_tier`.<br/><br/>When `useDefaultTierForNull` is `false`, the default value is `{}` i.e. no replicas to be loaded on any tier. |
|`useDefaultTierForNull`|Determines the default value of `tieredReplicants` if it is not specified or set to `null`.| No | `true`|

Specific types of load rules discussed below may have other properties too.

**Pull request merged**
